src/postgres.c

```diff
@@ -203,47 +175,17 @@ static int ignore_transction_blocks(char *stmt) {
   return 0;
 }
 
-
-/*
- * Check the result of our query.
- *
- * Don't need it for now.
- */
-/*
-static int check_result(PGresult *res, PGconn *conn, char *query) {
-  int code = PQresultStatus(res);
-
-  switch(code) {
-    case PGRES_TUPLES_OK:
-    case PGRES_COMMAND_OK: {
-      if (DEBUG)
-        printf("[Postgres] Executed: %s\n", query);
-      break;
-    }
-    default: {
-      char *err = PQerrorMessage(conn);
-      printf("[%d] Error: %s\n", code, err);
-      break;
-    }
-  }
-
-  PQclear(res);
-
-  return code;
-}
-*/
-
 /*
  * Shutdown pool.
  */
-void postgres_free() {
+void postgres_free(void) {
   int i;
 
-  /* Signal the workers to shut down */
-  shutdown = 1;
   for (i = 0; i < POOL_SIZE; i++) {
-    pthread_mutex_unlock(&signals[i]);
-    pthread_join(threads[i], NULL);
+    /* Kill */
+    pthread_cancel(threads[i]);
```

Comment:
This is racy. `pthread_cancel` is async and isn't even guaranteed to succeed. A clean approach would be to close the write end of the pipe. The workers would then read a `0` length. Then they can return from the loop and clean up their connection.

The downside of that is that they will drain the queue first which isn't optimal if you are in a hurry. If you are in a hurry you can have a `_Atomic _Bool still_running` flag that they check before sending each statement or kill and join them.